src/ProvidedTypes.fs

```diff
@@ -6645,25 +6646,34 @@ namespace ProviderImplementation.ProvidedTypes.AssemblyReader
             with err -> 
               failwithf  "FAILED decodeILCustomAttribData, data.Length = %d, data = %A, meth = %A, argtypes = %A, fixedArgs=%A, nnamed = %A, sigptr before named = %A,  innerError = %A" bytes.Length bytes ca.Method.EnclosingType ca.Method.FormalArgTypes fixedArgs nnamed sigptr (err.ToString())
 
-        type CacheValue = ILModuleReader
-        let (|CacheValue|_|) (wr: WeakReference) = match wr.Target with null -> None | v -> Some (v :?> CacheValue)
-        let CacheValue (reader: CacheValue) = System.WeakReference reader
+        // Share DLLs across providers by caching them
+        let readerCache = ConcurrentDictionary<(string * string), DateTime * int * ILModuleReader>()
 
-        // Amortize readers weakly - this is enough that all the type providers in this DLL will at least share
-        // resources when all instantiated at the same time.
-        let readersWeakCache = ConcurrentDictionary<(string * string), WeakReference>()
+        let createReader ilGlobals file =
+            let bytes = File.ReadAllBytes(file) // this will be on LOH, but it is long lived and not changing frequently so it's ok.
+            let reader = ILModuleReader(file, ByteFile(bytes), ilGlobals, true)
+            reader
 
-        let ILModuleReaderAfterReadingAllBytes  (file:string, ilGlobals: ILGlobals) =
-            let bytes = File.ReadAllBytes file
+        let GetReaderCache () = ReadOnlyDictionary(readerCache)
+
+        let ILModuleReaderAfterReadingAllBytes (file:string, ilGlobals: ILGlobals) =
             let key = (file, ilGlobals.systemRuntimeScopeRef.QualifiedName)
-            match readersWeakCache.TryGetValue (key) with
-            | true, CacheValue mr2  when bytes = mr2.Bytes ->
-                mr2 // throw away the bytes we just read and recycle the existing ILModuleReader
-            | _ ->
-                let mr = ILModuleReader(file, ByteFile(bytes), ilGlobals, true)
-                readersWeakCache.[key] <- CacheValue (mr)
-                mr
+            let add = 
+                fun _ -> 
+                    let lastWriteTime = File.GetLastWriteTime(file)
+                    let reader = createReader ilGlobals file
+                    (lastWriteTime, 1, reader)
+            let update =
+                fun _ (currentLastWriteTime, count, reader) ->
+                    let lastWriteTime = File.GetLastWriteTime(file)
+                    if currentLastWriteTime <> lastWriteTime then
+                        let reader = createReader ilGlobals file
+                        (lastWriteTime, count + 1, reader)
+                    else
+                        (lastWriteTime, count, reader)
 

```

Comment:
When do you remove?

Reply:
How about a MemoryCache with a 30 second expiration?

Reply:
Timed caches I guess can work, but I still feel a little uncomfortable with it.